pkg/migrations/00004_add_blockchain_columns.up.sql

```diff
@@ -0,0 +1,12 @@
+CREATE TABLE blockchain_messages(
+	block_number BIGINT NOT NULL,
+	block_hash BYTEA NOT NULL,
+	originator_node_id INT NOT NULL,
+	originator_sequence_id BIGINT NOT NULL,
+	is_canonical BOOLEAN NOT NULL DEFAULT TRUE,
+	PRIMARY KEY (block_number, block_hash, originator_node_id, originator_sequence_id),
+	FOREIGN KEY (originator_node_id, originator_sequence_id) REFERENCES gateway_envelopes(originator_node_id, originator_sequence_id)
+);
+
+CREATE INDEX idx_blockchain_messages_block_canonical ON blockchain_messages(block_number, is_canonical);
```

Comment:
the `GetBlocksInRange` can't actually use this index because it is missing parts of the query. You need a composite index such as:
```
CREATE INDEX idx_blockchain_messages_block_hash_canonical 
ON blockchain_messages(block_number, is_canonical, block_hash);
```

Reply:
maybe the index without the hash is sufficient and the table scans won't be too bad. Hard to predict without seeing a query plan.